File: app/src/main/java/com/raival/fileexplorer/d8/DiagnosticWrapper.java
package com.raival.fileexplorer.d8;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.util.Locale;
import java.util.Objects;

import javax.tools.Diagnostic;
import javax.tools.JavaFileObject;

public class DiagnosticWrapper implements Diagnostic<File> {
    public static final int USE_LINE_POS = -31;

    private String code;
    private File source;
    private Diagnostic.Kind kind;

    private long position;
    private long startPosition;
    private long endPosition;

    private long lineNumber;
    private long columnNumber;
    private View.OnClickListener onClickListener;
    private String message;

    /**
     * Extra information for this diagnostic
     */
    private Object mExtra;
    private int startLine;
    private int endLine;
    private int startColumn;
    private int endColumn;

    public DiagnosticWrapper() {

    }

    public DiagnosticWrapper(Diagnostic<? extends JavaFileObject> obj) {
        try {
            this.code = obj.getCode();
            if (obj.getSource() != null) {
                this.source = new File(obj.getSource().toUri());
            }
            this.kind = obj.getKind();

            this.position = obj.getPosition();
            this.startPosition = obj.getStartPosition();
            this.endPosition = obj.getEndPosition();

            this.lineNumber = obj.getLineNumber();
            this.columnNumber = obj.getColumnNumber();

            this.message = obj.getMessage(Locale.getDefault());
        } catch (Throwable e) {
            // ignored
        }
    }

    public View.OnClickListener getOnClickListener() {
        return onClickListener;
    }

    public void setOnClickListener(View.OnClickListener listener) {
        onClickListener = listener;
    }

    @Override
    public Kind getKind() {
        return kind;
    }

    public void setKind(Kind kind) {
        this.kind = kind;
    }

    @Override
    public File getSource() {
        return source;
    }

    public void setSource(File source) {
        this.source = source;
    }

    @Override
    public long getPosition() {
        return position;
    }

    public void setPosition(long position) {
        this.position = position;
    }

    @Override
    public long getStartPosition() {
        return startPosition;
    }

    public void setStartPosition(long startPosition) {
        this.startPosition = startPosition;
    }

    @Override
    public long getEndPosition() {
        return endPosition;
    }

    public void setEndPosition(long endPosition) {
        this.endPosition = endPosition;
    }

    @Override
    public long getLineNumber() {
        return lineNumber;
    }

    public void setLineNumber(long lineNumber) {
        this.lineNumber = lineNumber;
    }

    @Override
    public long getColumnNumber() {
        return columnNumber;
    }

    public void setColumnNumber(long columnNumber) {
        this.columnNumber = columnNumber;
    }

    @Override
    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.code = code;
    }

    @Override
    public String getMessage(Locale locale) {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public Object getExtra() {
        return mExtra;
    }

    public void setExtra(Object mExtra) {
        this.mExtra = mExtra;
    }

    @NonNull
    @Override
    public String toString() {
        return "startOffset: " + startPosition + "\n" +
                "endOffset: " + endPosition + "\n" +
                "position: " + position + "\n" +
                "startLine: " + startLine + "\n" +
                "startColumn: " + startColumn + "\n" +
                "endLine: " + endLine + "\n" +
                "endColumn: " + endColumn + "\n" +
                "message: " + message;
    }

    @Override
    public int hashCode() {
        return Objects.hash(code, source, kind, position, startPosition, endPosition, lineNumber,
                columnNumber, message, mExtra);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj instanceof DiagnosticWrapper) {
            DiagnosticWrapper that = (DiagnosticWrapper) obj;

            if (that.message != null && this.message == null) {
                return false;
            }

            if (that.message == null && this.message != null) {
                return false;
            }

            if (!Objects.equals(that.message, this.message)) {
                return false;
            }

            if (!Objects.equals(that.source, this.source)) {
                return false;
            }

            if (that.lineNumber != this.lineNumber) {
                return false;
            }

            return that.columnNumber == this.columnNumber;
        }
        return super.equals(obj);
    }

    public int getStartLine() {
        return startLine;
    }

    public void setStartLine(int line) {
        this.startLine = line;
    }

    public int getEndLine() {
        return endLine;
    }

    public void setEndLine(int endLine) {
        this.endLine = endLine;
    }

    public int getStartColumn() {
        return startColumn;
    }

    public void setStartColumn(int startColumn) {
        this.startColumn = startColumn;
    }

    public int getEndColumn() {
        return endColumn;
    }

    public void setEndColumn(int endColumn) {
        this.endColumn = endColumn;
    }
}

File: app/src/main/java/com/raival/fileexplorer/tab/file/d8/DiagnosticWrapper.kt
package com.raival.fileexplorer.tab.file.d8

import android.view.View
import java.io.File
import java.util.*
import javax.tools.Diagnostic

class DiagnosticWrapper : Diagnostic<File?> {
    private var code: String? = null
    private var source: File? = null
    private var kind: Diagnostic.Kind? = null
    private var position: Long = 0
    private var startPosition: Long = 0
    private var endPosition: Long = 0
    private var lineNumber: Long = 0
    private var columnNumber: Long = 0
    var onClickListener: View.OnClickListener? = null
    private var message: String? = null

    /**
     * Extra information for this diagnostic
     */
    private var extra: Any? = null
    private var startLine = 0
    private var endLine = 0
    private var startColumn = 0
    private var endColumn = 0

    override fun getKind(): Diagnostic.Kind {
        return kind!!
    }

    fun setKind(kind: Diagnostic.Kind) {
        this.kind = kind
    }

    override fun getSource(): File? {
        return source
    }

    fun setSource(source: File?) {
        this.source = source
    }

    override fun getPosition(): Long {
        return position
    }

    fun setPosition(position: Long) {
        this.position = position
    }

    override fun getStartPosition(): Long {
        return startPosition
    }

    fun setStartPosition(startPosition: Long) {
        this.startPosition = startPosition
    }

    override fun getEndPosition(): Long {
        return endPosition
    }

    fun setEndPosition(endPosition: Long) {
        this.endPosition = endPosition
    }

    override fun getLineNumber(): Long {
        return lineNumber
    }

    fun setLineNumber(lineNumber: Long) {
        this.lineNumber = lineNumber
    }

    override fun getColumnNumber(): Long {
        return columnNumber
    }

    fun setColumnNumber(columnNumber: Long) {
        this.columnNumber = columnNumber
    }

    override fun getCode(): String {
        return code!!
    }

    fun setCode(code: String?) {
        this.code = code
    }

    override fun getMessage(locale: Locale): String {
        return message!!
    }

    fun setMessage(message: String?) {
        this.message = message
    }

    override fun toString(): String {
        return """
               startOffset: $startPosition
               endOffset: $endPosition
               position: $position
               startLine: $startLine
               startColumn: $startColumn
               endLine: $endLine
               endColumn: $endColumn
               message: $message
               """.trimIndent()
    }

    override fun hashCode(): Int {
        return Objects.hash(
            code, source, kind, position, startPosition, endPosition, lineNumber,
            columnNumber, message, extra
        )
    }

    override fun equals(other: Any?): Boolean {
        if (other is DiagnosticWrapper) {
            if (other.message != null && message == null) {
                return false
            }
            if (other.message == null && message != null) {
                return false
            }
            if (other.message != message) {
                return false
            }
            if (other.source != source) {
                return false
            }
            return if (other.lineNumber != lineNumber) {
                false
            } else other.columnNumber == columnNumber
        }
        return super.equals(other)
    }

    companion object {
        const val USE_LINE_POS = -31
    }
}